Make SPLIT_SHARED_LIBRARIES=OFF more robust

azat opened the pull request:
Refs: #6915
Closes: #9006
Fixes: #9005

Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Make SPLIT_SHARED_LIBRARIES=OFF more robust

azat committed:
Make SPLIT_SHARED_LIBRARIES=OFF more robust

Refs: #6915
Closes: #9006
Fixes: #9005

alexey-milovidov reviewed (approved):
I have added a comment with explanation. Please check if it's correct.

azat commented:
>I have added a comment with explanation. Please check if it's correct.

I think it is, thanks!

azat committed:
Use -undefined,dynamic_lookup over --unresolved-symbols=ignore-all on OSX

Since OSX does not have later [1]:

    ld: unknown option: --unresolved-symbols=ignore-all
    clang: error: linker command failed with exit code 1 (use -v to see invocation)

  [1]: https://clickhouse-builds.s3.yandex.net/9156/eb89764d9dd631a9227ad9c36ad7d6f3ac8d0a4e/build_log_611498641_1581938330.txt

azat commented:
>ClickHouse build check — 16/16 builds are OK

Builds are ok

alexey-milovidov commented on dbms/CMakeLists.txt:
```diff
@@ -268,6 +268,14 @@ endif ()
 add_subdirectory(src/Common/ZooKeeper)
 add_subdirectory(src/Common/Config)
 
+function(target_ignore_unresolved_symbols target)
```
Where is my comment? :(

alexey-milovidov commented:
Sorry, it doesn't work as expected. I'm always using static linking. And now the linker doesn't show me errors. Instead, I get errors at startup time:

```
milovidov@milovidov-Pro-P30:~/work/ClickHouse/dbms/programs/server$ ./clickhouse-server
Processing configuration file 'config.xml'.
Merging configuration file 'config.d/log_to_console.xml'.
Merging configuration file 'config.d/macros.xml'.
Merging configuration file 'config.d/metric_log.xml'.
Merging configuration file 'config.d/more_clusters.xml'.
Merging configuration file 'config.d/part_log.xml'.
Merging configuration file 'config.d/path.xml'.
Merging configuration file 'config.d/query_masking_rules.xml'.
Merging configuration file 'config.d/text_log.xml'.
Merging configuration file 'config.d/zookeeper.xml'.
Include not found: clickhouse_remote_servers
Include not found: clickhouse_compression
Logging trace to console
./clickhouse-server: symbol lookup error: ./clickhouse-server: undefined symbol: _Z8setColorm
```

azat commented:
>Sorry, it doesn't work as expected.

It does actually, the problem is that expectation is wrong.

> I'm always using static linking. And now the linker doesn't show me errors. Instead, I get errors at startup time:

I don't use static linking (shard + split + unbundled), and using this setup I got unresolved symbols at startup time too, that said that `-Wl,--unresolved-symbols=ignore-all` just hides the problem, and while you reverted one usage there is [one left](https://github.com/ClickHouse/ClickHouse/blob/c647da3cf5aff84695afbc8c316b45c4045f2e81/dbms/CMakeLists.txt#L269).

So that said that right now `-Wl,--unresolved-symbols=ignore-all` usage will just hide errors.

P.S. I assumed that it will just hide the problems but did not tested this (my bad), since `-Wl,--unresolved-symbols=ignore-all`  was already used for shared build.
